src/jdk.httpserver/share/classes/sun/net/httpserver/ServerImpl.java

```diff
@@ -374,10 +374,16 @@ public void run() {
 
                     /* process the selected list now  */
                     Set<SelectionKey> selected = selector.selectedKeys();
-                    Iterator<SelectionKey> iter = selected.iterator();
-                    while (iter.hasNext()) {
-                        SelectionKey key = iter.next();
-                        iter.remove ();
+                    // create a copy of the selected keys so that we can iterate over it
+                    // and at the same time not worry about the underlying Set being
+                    // modified (leading to ConcurrentModificationException) due to
+                    // any subsequent select operations that we invoke on the
+                    // selector (in this same thread).
+                    final Set<SelectionKey> copy = new HashSet<>(selected);
+                    // iterate over the copy
+                    for (final SelectionKey key : copy) {
```

Comment:
Another possibility would be to call toArray() - since we're simply going to iterate we don't need a full copy of the hashset - e.g.: `for (var key : selected.toArray(SelectionKey[]::new)) {`

Reply:
That's a good idea. I've now updated the PR to use this suggestion.